### PR TITLE
Remote networks datasource

### PR DESCRIPTION
**Pull request by vmanilo**
Resolves https://github.com/Twingate/terraform-provider-twingate/issues/103

Based on PR: https://github.com/Twingate/terraform-provider-twingate/pull/112

## Changes
- added datasource for remote-networks

### CODE REVIEW

File: twingate/datasource_remote_networks.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	}
+
+	resourceData.SetId(networkName)
+
+	return diags
+}
+
+func datasourceRemoteNetworks() *schema.Resource {
+	return &schema.Resource{
+		Description: "A Remote Network represents a single private network in Twingate that can have one or more Connectors and Resources assigned to it. You must create a Remote Network before creating Resources and Connectors that belong to it. For more information, see Twingate's [documentation](https://docs.twingate.com/docs/remote-networks).",
```
Comment (vmanilo) [inline twingate/datasource_remote_networks.go:41]: @alexmensch please check text

File: twingate/datasource_remote_networks.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+		ReadContext: datasourceRemoteNetworksRead,
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "The name of the Remote Network",
```
Comment (vmanilo) [inline twingate/datasource_remote_networks.go:47]: @alexmensch please check text

File: twingate/datasource_remote_networks.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+			"remote_networks": {
+				Type:        schema.TypeList,
+				Computed:    true,
+				Description: "List of Remote Networks",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "The ID of the Remote Network",
```
Comment (vmanilo) [inline twingate/datasource_remote_networks.go:58]: @alexmensch please check text

File: twingate/datasource_remote_networks.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+						},
+						"name": {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "The name of the Remote Network",
```
Comment (vmanilo) [inline twingate/datasource_remote_networks.go:63]: @alexmensch please check text

**Review by alexmensch — APPROVED** [review]
Made a small addition to the text, otherwise all other text looks good to me.

**Review by ekampf — COMMENTED** [review]
One thing I'm missing is where in the docs does it explain what the data sources return?
The description for `remoote_network` and `remote_networks` is identical but one returns a single instance and one returns a list of matches...

**Comment by ekampf** [comment]
Never mind figured it out

File: twingate/client_resource.go [hunk 5]
```diff
@@ -13,12 +13,15 @@ import (
 const readResourceQueryGroupsSize = 50
 
 var (
-	ErrTooManyGroupsError        = errors.New("provider does not support more than 50 groups per resource")
-	ErrGraphqlIDIsEmpty          = errors.New("id is empty")
-	ErrGraphqlNameIsEmpty        = errors.New("name is empty")
-	ErrGraphqlConnectorIDIsEmpty = errors.New("connector id is empty")
-	ErrGraphqlNetworkIDIsEmpty   = errors.New("network id is empty")
-	ErrGraphqlNetworkNameIsEmpty = errors.New("network name is empty")
+	ErrTooManyGroupsError              = errors.New("provider does not support more than 50 groups per resource")
```
Comment (ekampf) [inline twingate/client_resource.go:16]: It doesn't seem like `client_resource` is the right place for these to be as these are generic to all clients...

Comment (ekampf) [reply twingate/client_resource.go:16]: Feel free to fix on a separate PR

**Review by ekampf — APPROVED** [review]
Approving but we need a followup PR to move clients t errors from client_resource.go to a better place

**Review by ekampf — CHANGES_REQUESTED** [review]
We're considering making Remote Network name unique which would mean we wouldn't need this data source and could query by either id or name in the non-plural one.

So lets put this PR on hold for now

File: docs/data-sources/remote_network.md [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+# twingate_remote_network (Data Source)
+
+A Remote Network represents a single private network in Twingate that can have one or more Connectors and Resources assigned to it. You must create a Remote Network before creating Resources and Connectors that belong to it. For more information, see Twingate's [documentation](https://docs.twingate.com/docs/remote-networks).
+
+## Example Usage
+
+```terraform
+data "twingate_remote_network" "foo" {
+  id = "<your network's id>"
```
Comment (alexmensch) [inline docs/data-sources/remote_network.md:17]: We should change this to the following to make it more explicit:

`id = "<remote network id>"`

And then we should clarify where this ID comes from, which I'll add in a separate comment.

File: docs/data-sources/remote_networks.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+# twingate_remote_networks (Data Source)
+
+A Remote Network represents a single private network in Twingate that can have one or more Connectors and Resources assigned to it. You must create a Remote Network before creating Resources and Connectors that belong to it. For more information, see Twingate's [documentation](https://docs.twingate.com/docs/remote-networks).
+
+## Example Usage
+
+```terraform
+data "twingate_remote_networks" "foo" {
+  name = "<your network's name>"
```
Comment (alexmensch) [inline docs/data-sources/remote_networks.md:17]: Change to:

`name = "<remote network name>"`

File: docs/data-sources/remote_network.md [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+}
+```
+
```
Comment (alexmensch) [inline docs/data-sources/remote_network.md:20]: Let's add a clarification below this on why we have a plural data source:

`Remote Network names are not constrained to be unique within Twingate, so it is possible that this data source will return multiple list items.`

**Review by alexmensch — CHANGES_REQUESTED** [review]

**Comment by ekampf** [comment]
Closing as we now support name uniqueness and this is unecessary

**Pull request closed without merging**